Core: Auto promote sorted column metrics to truncate(16)

szehon-ho opened the pull request:
-- Add flag "write.metadata.sorted.metrics.default" which defaults to "full"
-- Fix unrelated bug in configuring nested Orc Metrics
-- Welcome feedback how to avoid propagation of sortOrder through AppenderFactories

szehon-ho commented:
PR Background:

A suggestion from @aokolnychyi , we find that users do not tend to fine-tune column metrics.  The default truncate(16) does not do much for certain sortable types, and reduces benefits of predicate pruning.  Cost of storing full metrics should be small as number of sorted columns will be generally small.

aokolnychyi commented:
Thanks for the PR, @szehon-ho! Let me take a look now.

aokolnychyi commented on core/src/main/java/org/apache/iceberg/TableProperties.java:
```diff
@@ -120,6 +120,8 @@ private TableProperties() {
   public static final String METRICS_MODE_COLUMN_CONF_PREFIX = "write.metadata.metrics.column.";
   public static final String DEFAULT_WRITE_METRICS_MODE = "write.metadata.metrics.default";
   public static final String DEFAULT_WRITE_METRICS_MODE_DEFAULT = "truncate(16)";
+  public static final String SORTED_COL_DEFAULT_METRICS_MODE = "write.metadata.sorted.metrics.default";
```
Actually, I think we should simplify it a bit. The use case I was talking about is when the user configures the default sort mode as `none` or `counts` but creates a table with a sort order. In that case, we should promote the metrics for sort columns to be at least `truncate(16)` unless the user sets a mode for sort columns explicitly. It is probably too dangerous to promote to `full` as the values may be too long. I guess `truncate(16)` is a reasonable default for sort columns and it will apply only to sort and binary columns. Longs/integers are not affected.

Internally, we may want to change the default value to `counts` instead of `truncate(16)` for tables with many columns as we have a lot of tables with 100+ columns and we don't want to a ton of unnecessary metadata. But I am not sure that the community wants to do the same.

aokolnychyi replied:
Thoughts, @szehon-ho @RussellSpitzer @rdblue?

aokolnychyi replied:
I'd also say the promotion should be implicit and we don't need extra table properties.

RussellSpitzer replied:
I think i'm also in favor of just making the change without a flag to configure. Just to make configuration have one parameter less, do we really have a lot of use cases where truncate(16) isn't the right thing to do? And if it isn't we can just let users manually reconfigure those columns individually if they need.

aokolnychyi replied:
@szehon-ho, could you update the PR to match the behavior in [this](https://github.com/apache/iceberg/pull/2240#discussion_r577899216) comment, please?

szehon-ho replied:
Yes I'll take a look, thanks for the feedback guys

szehon-ho replied:
Recap if I understand correctly: if default < truncate(16), then sort columns => truncate(16).  And no table property (user overrides the sort col metric property itself)

RussellSpitzer replied:
That’s just my suggestion, I can always be convinced otherwise too 😊

szehon-ho replied:
No you guys have a point, full is unnecessary if someone sorts by a huge text for example, and there's already a good way to override the config.  Updated the PR and changed the title to reflect.

rdblue replied:
I agree with this behavior and I also would not add a new table property.

aokolnychyi commented:
Let me do another round now.

aokolnychyi commented:
I see that the current approach tries to promote the sort columns on read, which is one way of doing this. I think I'd prefer the approach we took with `validateReferencedColumns` that does its validation during table creation and in a few other relevant places. If we do it that way, we don't need to modify any ORC and Parquet code and the scope of the change will be really small.

Basically, we could modify `TableMetadata#newTableMetadata` as this:

```
    MetricsConfig metricsConfig = MetricsConfig.fromProperties(properties);
    metricsConfig.validateReferencedColumns(schema);
    MetricsConfig freshMetricsConfig = metricsConfig.copyWithPromotedSortColumns(freshSortOrder);

    Map<String, String> freshProperties = Maps.newHashMap(properties);
    freshProperties.putAll(freshMetricsConfig.toProperties());
```

That will need two new methods in `MetricsConfig` (let's not make them public):

```
  MetricsConfig copyWithPromotedSortColumns(SortOrder sortOrder) {
    // create a copy
    // check the default mode
    // if none or count -> iterate through the sort order fields and promote as needed if not set by the user
  }

  Map<String, String> toProperties() {
    // map into table props
  }
```

What are your thoughts, @RussellSpitzer @rdblue @holdenk @shardulm94 @pvary?

RussellSpitzer commented:
Would that also catch alters? @aokolnychyi? I don't think that's unreasonable since it does make sense to keep it off the read path if possible ...

aokolnychyi commented:
We will catch altering sort order separately just like we catch a few other places for `validateReferencedColumns`.

rdblue commented:
I don't think I would approach the problem as @aokolnychyi is suggesting. I thought that this would promote when accessing the metrics. So basically `MetricsConfig.fromProperties` becomes `MetricsConfig.forTable` so that it can be based on both the properties and the sort order for a table. I think that would be a reasonably small change.

The advantage of that approach is that the sort order doesn't make changes to table properties. Those are always configured by users.

aokolnychyi commented:
> So basically MetricsConfig.fromProperties becomes MetricsConfig.forTable so that it can be based on both the properties and the sort order for a table.

The problem is that we build `MetricsConfig` in places where we don't have access to the table object. We made a recent change to make `BaseTable` serializable but we would need to change quite some places to pass it around. Also, I am not sure about a potential performance penalty. We do serialize things like spec and props so maybe it won't be a big deal to serialize the complete table object but someone has to test that. We could pass the sort order around but it anyway seemed like a major work to redo that in all places. Plus, it is easy to miss something. For example, the current change does not capture Flink but all tests are green.

> The advantage of that approach is that the sort order doesn't make changes to table properties. Those are always configured by users.

I think having something like `toTableProperties` on `MetricsConfig` will make it possible to migrate to column ids instead of names in the future. If we want to use column ids, then modifying table props seems inevitable to me.

aokolnychyi commented:
But I do get the concern of modifying table properties. However, it looks there will be a few places where will need to do that (e.g. schema evolution).

rdblue commented:
My concern isn't modifying table properties specifically. We need to do that so that we update the properties to reflect the user's intent. If we have a setting for a column, renaming the property to carry it through a column rename is preserving the user's intent. But if we update table properties then we are potentially losing what the user chose to do.

For example, what do we do when the sort field is removed and there is a metrics config property? We can't remove a `truncate[16]` setting because we don't know whether it was added when the field was added to the sort, or if it was specifically set by the table owner. Similarly, we have cases where the owner may set a field to `counts` and then add it to the sort. I think that the table owner's config should take precedence. So we don't replace the property, but how then does the owner signal that it should be defaulted instead? Last, how do we handle tables that currently have sort orders? If we base metrics on the current sort order then everything starts working automatically. If we don't, then we would need to trigger a commit to update the properties.

Even if we don't pass the table around, I think we should definitely pass the sort order in to get the metrics config. It is probably not the time to refactor and pass the table, but we can pass the sort order like we do the other table config.

aokolnychyi commented:
> For example, what do we do when the sort field is removed and there is a metrics config property?

Yeah, that's the point we won't be able to differentiate if we modified the props. I am convinced now.
If we have enough confidence that this feature will be useful, I am ok if we want to modify all places where we write.

I took a look at how hard will it be to move to `Table` as that would simplify our life in the future. Since `BaseTable` uses a proxy for serializing, we only serialize `FileIO` and the metadata location. We will need to read the metadata file back on each executor. In addition, `FileIO` may contain a full Hadoop conf. That's why we broadcast `FileIO` and `EncryptionManager` in Spark. We could consider broadcasting `Table` instead of `FileIO` and `EncryptionManager`. However, we will still have to read the metadata file from each executor. Is that a big deal?

I am fine either way.

rdblue commented:
It sounds like it would take some work to serialize Table. I think we'd want to update it so that at least one version of TableMetadata is also carried through to avoid that read on every executor. In the long run, that's probably a good thing to do. But for now we can make easier progress serializing things separately.

Eventually, I think we should broadcast the table and FileIO together to take care of this issue for all of the things that might be expensive to serialize.

szehon-ho commented:
Thanks guys for the time and discussion (learning a lot on the side).

If I understand correctly, it looks metrics-promotion on read is the way to go due to simplicity, esp in regards to alter table not having to alter auto-promoted metrics properties.  But in this approach we have to manually pass in sortOrder in all places of write, and I missed it in the new SparkTableUtil and Flink, which I can add.

I am wondering (being a better code aside), for this particular change what's the use-case covered by passing SortOrder instead of the latest table's sorted column names , auto-promotion of sorted columns would work given just the current sorted names, or no?  I agree SortOrder is less hacky, but just to check my understanding, thanks,

szehon-ho commented:
Edit: and for now as Table is hard to serialize, we can have MetricsConfig.forSortOrder(properties, sortOrder), to make progress, as SortOrder is already serializable.

szehon-ho commented:
Hi @rdblue @aokolnychyi  , so I propagated the SortOrder through the various writers as discussed, including the ones I missed before. 

I spent some time verifying it by unit tests (TestMergingMetrics gives good coverage for Flink/Generic/Spark AppenderFactory, and TestSparkDataWrite is an end-to-end test on Spark side).  When you get a chance, let me know what you think, thanks again.

aokolnychyi commented:
Sorry, I was distracted by other things. Let me take a look at this PR tomorrow.

aokolnychyi commented:
I think we have a similar effort in https://github.com/apache/iceberg/pull/2214.

szehon-ho commented:
Thanks for looking @aokolnychyi .  Oh yes what a coincidence, yea if that could be merged then I can rebase and just make the changes to the metrics part.

szehon-ho commented:
Rebased, and as discussed, make method MetricsConfig.fromTable(table), and propagate the table arg up the various writer stacks.  This was helped by changes of #2362 (table propagation to Spark) and #2499 (SparkAppenderFactory builder using table).  

This last change in particular aligns the other AppenderFactories (Flink/Generic) to also be constructed with tables, like SparkAppenderFactory.  Most of the other changes are test changes to accomodate this change of arguments to AppenderFactories to use Tables.

szehon-ho commented:
@aokolnychyi @RussellSpitzer  @karuppayya @flyrain @openinx for feedback, thank you

aokolnychyi commented:
Let me take a look later today. All prerequisite PRs are in, so we should be able to finish this one.

aokolnychyi commented on core/src/main/java/org/apache/iceberg/MetricsModes.java:
```diff
@@ -150,6 +150,19 @@ public String toString() {
+   * @param defaultMode default mode
+   * @return mode to use
+   */
+  public static MetricsMode promoteSortedColumnDefault(MetricsMode defaultMode) {
```
I'd probably move this into `MetricsConfig` as a private method. Seems like we only use it there.

szehon-ho replied:
Done

aokolnychyi commented:
I think this is the right direction. @szehon-ho, could you rebase this PR?

szehon-ho commented:
Thanks @aokolnychyi! rebased

aokolnychyi commented on core/src/main/java/org/apache/iceberg/MetricsConfig.java:
```diff
@@ -77,6 +80,14 @@ public static MetricsConfig getDefault() {
   }
 
   public static MetricsConfig fromProperties(Map<String, String> props) {
```
I've been thinking about this change while doing recent work around writer factories and deletes.

Naming-wise, I'd probably prefer calling it `forTable(table)`. If we decide to store [metrics for rows in position deletes](https://github.com/apache/iceberg/pull/2942#discussion_r683715471), we then can provide something like `forPositionDelete(table)` to promote the path and pos columns and apply the regular config for all other columns in optional `row`.

szehon-ho replied:
Renamed to 'forTable'

aokolnychyi commented on core/src/main/java/org/apache/iceberg/MetricsConfig.java:
```diff
@@ -87,6 +98,18 @@ public static MetricsConfig fromProperties(Map<String, String> props) {
     }
 
+    // First set sorted column with sorted column default (can be overridden by user)
+    final Set<String> sortedCols = new HashSet<>();
```
nit: we usually don't mark temp vars as final in Iceberg. It is a style preference. I know there were some performance optimizations in old JVMs for final vars but I think modern JVM are smart enough these days.

aokolnychyi replied:
You could probably init the set to the method call result instead of calling calling `addAll` later.

```
Set<String> sortColumns = order != null ? SortOrderUtil.sortColumns(order) : ImmutableSet.of();
...
```

aokolnychyi replied:
Or something like [this](https://github.com/apache/iceberg/pull/2240#discussion_r684560856). Up to you.

szehon-ho replied:
Yea true I dont usually put final on local variables, don't recall how that happened.  Anyway, simplified according to your suggestion.

aokolnychyi commented on core/src/main/java/org/apache/iceberg/util/SortOrderUtil.java:
```diff
@@ -64,4 +68,16 @@ static SortOrder buildSortOrder(Schema schema, PartitionSpec spec, SortOrder sor
   }
+
+  public static Set<String> getSortedColumns(SortOrder sortOrder) {
+    if (sortOrder == null) {
```
Do we need this extra check for null here? Looks like we validate the order is not null before calling. I'd probably simplify  and assume it is not null. It is a pretty internal utility class.

aokolnychyi replied:
Or keep the check here but remove in the caller. Up to you.

szehon-ho replied:
Done, removed in caller and kept null check in the method.

aokolnychyi commented on core/src/main/java/org/apache/iceberg/util/SortOrderUtil.java:
```diff
@@ -64,4 +68,16 @@ static SortOrder buildSortOrder(Schema schema, PartitionSpec spec, SortOrder sor
     return builder.build();
   }
+
+  public static Set<String> getSortedColumns(SortOrder sortOrder) {
```
nit: `getSortedColumns` -> `sortColumns`. We usually try to avoid `get` in method names. It is a style preference we follow in Iceberg. Such prefixes usually have no meaning and just make the names longer.

szehon-ho replied:
Yep, this was my first PR to iceberg and I realized this :).  Changed

aokolnychyi commented on core/src/main/java/org/apache/iceberg/MetricsConfig.java:
```diff
@@ -95,16 +118,32 @@ public static MetricsConfig fromProperties(Map<String, String> props) {
+   * @param defaultMode default mode
+   * @return mode to use
+   */
+  private static MetricsMode promoteSortedColumnDefault(MetricsMode defaultMode) {
```
nit: `promoteSortedColumnDefault` as just `sortColumnDefaultMode`?

szehon-ho replied:
done

aokolnychyi commented on core/src/main/java/org/apache/iceberg/MetricsConfig.java:
```diff
@@ -87,6 +98,18 @@ public static MetricsConfig fromProperties(Map<String, String> props) {
+    if (order != null) {
+      sortedCols.addAll(SortOrderUtil.getSortedColumns(order));
+      sortedCols.stream().forEach(sc -> {
+        if (!props.containsKey(METRICS_MODE_COLUMN_CONF_PREFIX + sc)) {
```
Do we need this if condition? I think it is fine to default the mode here. If the user sets an explicit value, it will override this one.

szehon-ho replied:
Done

aokolnychyi commented on core/src/main/java/org/apache/iceberg/MetricsConfig.java:
```diff
@@ -95,16 +118,32 @@ public static MetricsConfig fromProperties(Map<String, String> props) {
+            // Mode was invalid, log the error and use the default (or default for sorted columns)
             LOG.warn("Ignoring invalid metrics mode for column {}: {}", columnAlias, props.get(key), err);
-            mode = spec.defaultMode;
+            if (sortedCols.contains(columnAlias)) {
```
If you remove the if condition above, you will be able to avoid this change too.
I think something as below would work.

```
if (order != null) {
  MetricsMode sortColumnDefaultMode = sortColumnDefaultMode(defaultMode);
  Set<String> sortColumns = SortOrderUtil.sortColumns(order);
  for (String sortColumn : sortColumns) {
    spec.columnModes.put(sortColumn, sortColumnDefaultMode);
  }
}
```

szehon-ho replied:
The purpose of this if,  if it's an invalid metric for a sorted column, we pick the sorted default instead of the normal default.

It's probably over-engineering though?  I reverted this and edited the test that was asserting this

aokolnychyi commented on core/src/test/java/org/apache/iceberg/TestMetricUtil.java:
```diff
@@ -0,0 +1,63 @@
+   * @param schema table schema
+   * @return table
+   */
+  public static Table createTestTable(Map<String, String> properties, SortOrder sortOrder,
```
What about calling `table.replaceSortOrder()` within tests and using the existing `createTable`?

szehon-ho replied:
Yep that's cleaner, done

aokolnychyi commented on data/src/main/java/org/apache/iceberg/data/GenericAppenderFactory.java:
```diff
@@ -47,27 +47,31 @@
  */
 public class GenericAppenderFactory implements FileAppenderFactory<Record> {
```
We have just recently introduced `WriterFactory`. We will be moving to that API instead of appenders.
I think it will be sufficient to just cover the new API. Plus, it should be super simple as I thought about this change while implementing `BaseWriterFactory`. It already has a reference to `Table`. This will help us to avoid changing so many places around `AppenderFactory` that we will remove in the nearest future.

szehon-ho replied:
Done, removed changes to unnecessary files

aokolnychyi reviewed (commented):
Sorry for the delay, @szehon-ho! I waited for a couple of other changes to have a clear picture. I did one pass and had only cosmetic comments.

I'd probably consume the new logic only in `BaseWriterFactory`, our new API that will replace `FileAppenderFactory`. Should make your life easier and make the scope of this PR smaller.

szehon-ho commented:
@aokolnychyi No problem, thanks for the review, addressed the comments and implemented only for new WriterFactory, when you are ready for another look.
  
By the way, I was thinking to move out the ORC nested column metric fix to a separate patch, what do you think?

szehon-ho commented:
Split out ORC nested metric mode fix out in a new PR : https://github.com/apache/iceberg/pull/2977 , which should make this change even smaller

aokolnychyi commented:
Thanks, @szehon-ho! Let me take a look now.

aokolnychyi commented on data/src/test/java/org/apache/iceberg/io/TestWriterMetrics.java:
```diff
@@ -0,0 +1,150 @@
+    this.fileFormat = fileFormat;
+  }
+
+  protected abstract WriterFactory newWriterFactory(Schema dataSchema);
```
nit: parameterize it as `WriterFactory<T>`?

aokolnychyi commented on data/src/test/java/org/apache/iceberg/io/TestWriterMetrics.java:
```diff
@@ -0,0 +1,150 @@
+  @Test
+  public void verifySortedColMetric() throws Exception {
+    T row = toRow(3, "3", true, 3L);
+    FileAppender<T> appender = newWriterFactory(SCHEMA).newDataWriter(
```
I think you can use `DataWriter#toDataFile` to get accesses to bounds and avoid exposing the appender.

szehon-ho replied:
Done

aokolnychyi commented on parquet/src/main/java/org/apache/iceberg/parquet/Parquet.java:
```diff
@@ -125,7 +125,7 @@ private WriteBuilder(OutputFile file) {
       schema(table.schema());
       setAll(table.properties());
-      metricsConfig(MetricsConfig.fromProperties(table.properties()));
+      metricsConfig(MetricsConfig.forTable(table));
```
Shall we do the same for ORC and Avro?

szehon-ho replied:
Changed

aokolnychyi reviewed (commented):
Looks like this one is converging. I did one more pass.

Let's get #2977 in and then rebase this one.

aokolnychyi commented on core/src/main/java/org/apache/iceberg/util/SortOrderUtil.java:
```diff
@@ -64,4 +68,16 @@ static SortOrder buildSortOrder(Schema schema, PartitionSpec spec, SortOrder sor
+    } else {
+      return sortOrder.fields().stream()
+          .map(SortField::sourceId)
+          .map(sid -> sortOrder.schema().findColumnName(sid))
```
After another look with fresh eyes, I don't think my statement above is accurate. If we sort using an order preserving transform, it means the source columns are somehow sorted (but may not be perfectly sorted).

For example, sorting by `month(date)`, means our dates are sorted across months but not necessarily within each month. Still, it probably makes sense to promote such columns. Identity and truncate transforms are also order preserving.

@RussellSpitzer @jackye1995, thoughts?

szehon-ho replied:
Yea it makes sense.  I wonder a bit the use case for sorting by bucket(id, 8), but yea it does not make too much sense to auto-promote them.

szehon-ho replied:
Yea makes sense.  I wonder a bit the use case for sorting by bucket(id), but in that case it makes sense to not promote it.

aokolnychyi replied:
The implementation here looks correct to me now.

aokolnychyi commented:
@szehon-ho, the other PR is in. Could you rebase this one?

szehon-ho commented:
Rebased and still working through some of the review comments.  Need to take a bit of look at the order preserving transform.

szehon-ho commented on core/src/test/java/org/apache/iceberg/TestMetrics.java:
```diff
@@ -106,9 +113,14 @@
   private static final Record FLOAT_DOUBLE_RECORD_1 = createRecordWithFloatAndDouble(1.2F, 3.4D);
   private static final Record FLOAT_DOUBLE_RECORD_2 = createRecordWithFloatAndDouble(5.6F, 7.8D);
   private static final Record NAN_ONLY_RECORD = createRecordWithFloatAndDouble(Float.NaN, Double.NaN);
-
```
Added separation

szehon-ho commented on core/src/test/java/org/apache/iceberg/TestMetrics.java:
```diff
@@ -597,6 +609,101 @@ public void testTruncateBinaryMetricsMode() throws IOException {
+    secondRecord.setField("fixedCol", fixed);
+    secondRecord.setField("binaryCol", ByteBuffer.wrap("S".getBytes()));
+    secondRecord.setField("timestampColBelowEpoch", DateTimeUtil.timestampFromMicros(0L));
+
```
Fixed

szehon-ho commented on data/src/main/java/org/apache/iceberg/data/BaseWriterFactory.java:
```diff
@@ -85,7 +85,7 @@ protected BaseWriterFactory(Table table, FileFormat dataFileFormat, Schema dataS
     EncryptionKeyMetadata keyMetadata = file.keyMetadata();
     Map<String, String> properties = table.properties();
-    MetricsConfig metricsConfig = MetricsConfig.fromProperties(properties);
+    MetricsConfig metricsConfig = MetricsConfig.forTable(table);
```
Updated

szehon-ho commented on data/src/test/java/org/apache/iceberg/io/TestWriterMetrics.java:
```diff
@@ -0,0 +1,150 @@
+
+  protected abstract WriterFactory newWriterFactory(Schema dataSchema);
+
+
```
Fixed

szehon-ho replied:
Done

szehon-ho commented on core/src/main/java/org/apache/iceberg/MetricsConfig.java:
```diff
@@ -101,10 +116,22 @@ public static MetricsConfig fromProperties(Map<String, String> props) {
           }
           spec.columnModes.put(columnAlias, mode);
         });
-
```
Done

szehon-ho commented on core/src/main/java/org/apache/iceberg/io/DataWriter.java:
```diff
@@ -86,4 +86,8 @@ public DataFile toDataFile() {
     return dataFile;
   }
+
+  public FileAppender<T> appender() {
```
Done

szehon-ho commented on core/src/main/java/org/apache/iceberg/MetricsConfig.java:
```diff
@@ -87,6 +97,11 @@ public static MetricsConfig fromProperties(Map<String, String> props) {
+    // First set sorted column with sorted column default (can be overridden by user)
+    MetricsMode sortedColDefaultMode = sortColumnDefaultMode(spec.defaultMode);
+    Set<String> sortedCols = SortOrderUtil.sortedColumns(order);
+    sortedCols.stream().forEach(sc -> spec.columnModes.put(sc, sortedColDefaultMode));
```
Renamed method to sortedColumnDefaultValue

szehon-ho commented on core/src/main/java/org/apache/iceberg/MetricsConfig.java:
```diff
@@ -87,6 +97,11 @@ public static MetricsConfig fromProperties(Map<String, String> props) {
 
+    // First set sorted column with sorted column default (can be overridden by user)
+    MetricsMode sortedColDefaultMode = sortColumnDefaultMode(spec.defaultMode);
+    Set<String> sortedCols = SortOrderUtil.sortedColumns(order);
```
Renamed method to sortedColumnDefaultMode

aokolnychyi replied:
Looks consistent now!

szehon-ho commented on core/src/test/java/org/apache/iceberg/TestSortOrder.java:
```diff
@@ -315,4 +318,15 @@ public void testEmptySortOrder() {
   }
+
+  @Test
+  public void testSortOrderColumnNames() {
```
Done

szehon-ho commented on core/src/test/java/org/apache/iceberg/TestMetrics.java:
```diff
@@ -597,6 +609,101 @@ public void testTruncateBinaryMetricsMode() throws IOException {
+        .asc("dateCol").build();
+
+    Table table =
+        TestTables.create(tableDir, "test", SIMPLE_SCHEMA, PartitionSpec.unpartitioned(), sortOrder, FORMAT_V2);
```
Added a spec which is PartitionSpec.unpartitioned(), if that is what you meant

szehon-ho commented on core/src/test/java/org/apache/iceberg/TestMetricsModes.java:
```diff
@@ -19,22 +19,39 @@
+    TestTables.clearTables();
+  }
+
+
```
Fixed

szehon-ho commented on data/src/test/java/org/apache/iceberg/io/TestWriterMetrics.java:
```diff
@@ -0,0 +1,150 @@
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+
+
```
Done

szehon-ho commented:
@aokolnychyi  the comments should be addressed now when you have time for another look, thanks again

rdblue commented on core/src/test/java/org/apache/iceberg/TestMetrics.java:
```diff
@@ -613,6 +626,99 @@ public void testTruncateBinaryMetricsMode() throws IOException {
+        .asc("stringCol")
+        .asc("dateCol").build();
+    PartitionSpec spec = PartitionSpec.unpartitioned();
+    Table table = TestTables.create(tableDir, "test", SIMPLE_SCHEMA, spec, sortOrder, FORMAT_V2);
```
Why does this test create a v2 table? Won't this work with v1 as well?

szehon-ho replied:
Originally I wanted to not increase test runtime that much, as metrics seemed pretty independent of formatVersion.

Changed to make the tests parameterized to run with v1 and v2, or let me know if prefer the original.

rdblue commented on core/src/main/java/org/apache/iceberg/MetricsConfig.java:
```diff
@@ -77,6 +79,14 @@ public static MetricsConfig getDefault() {
 
   public static MetricsConfig fromProperties(Map<String, String> props) {
+    return from(props, null);
+  }
```
Should we deprecate `fromProperties` or is there still a reasonable use for it? Looks like we're moving over all of our internal calls to `forTable`.

szehon-ho replied:
Done

aokolnychyi replied:
I'll resolve this as it has been addressed.

rdblue reviewed (approved):
I left a couple comments, but overall I think this looks ready. @aokolnychyi can you take another look?

karuppayya commented on core/src/main/java/org/apache/iceberg/MetricsConfig.java:
```diff
@@ -77,6 +79,14 @@ public static MetricsConfig getDefault() {
+    return from(props, null);
+  }
+
+  public static MetricsConfig forTable(Table table) {
```
Should we also use this method at `org.apache.iceberg.parquet.Parquet.WriteBuilder#forTable`, since we have replaced other usages in Parquet.java? https://github.com/apache/iceberg/pull/2240/files#diff-8b42edd2a56abae1dc480f36b0e8126c3742d4f41fd7db8ca0df358527940c4cR394

szehon-ho replied:
Good catch, done

aokolnychyi replied:
Resolving as this seems to be done.

karuppayya commented on core/src/test/java/org/apache/iceberg/TestMetricsModes.java:
```diff
@@ -19,22 +19,38 @@
+  @Rule
+  public TemporaryFolder temp = new TemporaryFolder();
+
+  private static final int FORMAT_V2 = 2;
```
nit: Should we have it defined as public constant in a utils files, so that it can be referenced in future?

szehon-ho replied:
I actually made this a parameter as per the comment above , but if we don't do it that way I can look at this.

karuppayya reviewed (commented):
Left few minor comments.Thanks @szehon-ho for working on this

szehon-ho commented on data/src/test/java/org/apache/iceberg/io/TestWriterMetrics.java:
```diff
@@ -0,0 +1,150 @@
+  public static Object[][] parameters() {
+    return new Object[][] {
+        {FileFormat.ORC},
+        {FileFormat.PARQUET}
```
Avro returns null metrics, so it seems usually skipped from Metrics tests. 
 https://github.com/apache/iceberg/blob/master/core/src/main/java/org/apache/iceberg/avro/AvroMetrics.java

szehon-ho commented on data/src/test/java/org/apache/iceberg/io/TestWriterMetrics.java:
```diff
@@ -0,0 +1,150 @@
+
+  @Before
+  public void setupTable() throws Exception {
+    this.tableDir = temp.newFolder();
```
Done

szehon-ho commented:
@rdblue and @karuppayya thanks for taking a look!  Addressed the comments, hope I understood them correctly.  For the original format V2 hardcoding in the test, I parameterize the new tests with v1 and v2, let me know if that's not what we want to do.

aokolnychyi commented:
Sorry, I was off last week. I'll take a look now.

aokolnychyi commented on core/src/main/java/org/apache/iceberg/MetricsConfig.java:
```diff
@@ -87,6 +107,11 @@ public static MetricsConfig fromProperties(Map<String, String> props) {
     }
 
+    // First set sorted column with sorted column default (can be overridden by user)
+    MetricsMode sortedColDefaultMode = sortedColumnDefaultMode(spec.defaultMode);
```
Looks clean!

aokolnychyi reviewed (approved):
Looks solid to me too. I think we need to use the new method in `WriteBuilder` in Avro.
I'd also slightly tweak the javadoc.

szehon-ho commented on core/src/main/java/org/apache/iceberg/MetricsConfig.java:
```diff
@@ -76,7 +78,25 @@ public static MetricsConfig getDefault() {
     }
   }
 
+  /**
```
Done and fixed case sensitivity.  I also removed the return type documentation as it's now a bit redundant with this javadoc, let me know you prefer to keep it.

szehon-ho commented on core/src/main/java/org/apache/iceberg/avro/Avro.java:
```diff
@@ -256,7 +256,7 @@ public DataWriteBuilder forTable(Table table) {
       withSpec(table.spec());
       setAll(table.properties());
-      metricsConfig(MetricsConfig.fromProperties(table.properties()));
+      metricsConfig(MetricsConfig.forTable(table));
```
Done

aokolnychyi commented:
Looks great to me. Thanks for the work, @szehon-ho! I'll merge when tests pass.
